### PR TITLE
NAS-119668: Correctly report vdev width

### PR DESCRIPTION
**Pull request by dszidi**
To test:

Create each configuration and check the reported width in the Storage Dashboard topology card.

- Two raidz1 vdevs with 5 disks in each vdev. Topology card should say "5 wide"
- Two mirror vdevs with 2 disks in each vdev. Topology card should say "2 wide"
- Two mirrors. One made with two disks and another with 3 disks. Topology card should say "Mixed Width VDEVs"

### CODE REVIEW

File: src/app/pages/storage/components/dashboard-pool/topology-card/topology-card.component.spec.ts [hunk 1]
```diff
@@ -35,31 +35,45 @@ describe('TopologyCardComponent', () => {
             data: [
               {
                 type: CreateVdevLayout.Raidz1,
-                children: [{ type: 'DISK', disk: 'sda' }, { type: 'DISK', disk: 'sdb' }, { type: 'DISK', disk: 'sdc' }],
+                children: [
+                  { type: 'DISK', disk: 'sda' },
```
Comment (undsoft) [inline src/app/pages/storage/components/dashboard-pool/topology-card/topology-card.component.spec.ts:39]: While we are here, can we please change 'DISK' to TopologyItemType.Disk?

**Review by undsoft — APPROVED** [review]
Looks good.
Can we extract this code in some utility function and add more unit tests for every scenario?
It may come in handy in new pool creation wizard.

File: src/app/pages/storage/components/dashboard-pool/topology-card/topology-card.component.ts [hunk 2]
```diff
@@ -20,7 +23,7 @@ interface TopologyState {
 }
 
 const notAssignedDev = 'VDEVs not assigned';
```
Comment (undsoft) [inline src/app/pages/storage/components/dashboard-pool/topology-card/topology-card.component.ts:25]: Feels weird comparing to a string down below in code. Should be something else and string itself should be translated.

File: src/app/services/storage.service.ts [hunk 3]
```diff
@@ -386,4 +397,149 @@ export class StorageService {
...
+      vdevTypes.add(vdev.type);
+    });
+    return vdevTypes;
+  }
+
+  isMixedVdevType(vdevTypes: Set<string>): boolean {
+    return vdevTypes.size > 1;
+  }
+
+  validateVdevs(category: PoolTopologyCategory,
```
Comment (undsoft) [inline src/app/services/storage.service.ts:487]: I believe you mentioned that there were some scenarios where we may want to compare things between different categories. In this case this method may not work.
You may want to make other methods that are not intended to be used externally `private` and extensively test public methods.

File: src/app/core/testing/utils/mock-storage-generator.utils.ts [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+      status: 'ONLINE',
+      topology: {
+        data: [],
+        special: [],
+        log: [],
+        spare: [],
+        cache: [],
+        dedup: [],
+      },
+    } as unknown as PoolInstance;
```
Comment (undsoft) [inline src/app/core/testing/utils/mock-storage-generator.utils.ts:59]: unknown appears to be unnecessary and you can use enum for `status`.

File: src/app/core/testing/utils/mock-storage-generator.utils.ts [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+    const disks: StorageDashboardDisk[] = [];
+
+    return { poolState: pool, disks };
+  }
+
+  addDataTopology(scenario: MockStorageScenario,
```
Comment (undsoft) [inline src/app/core/testing/utils/mock-storage-generator.utils.ts:66]: If you have more than 3 arguments, it's better to just have one argument called `options`.

File: src/app/core/testing/utils/mock-storage-generator.utils.ts [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+      case TopologyItemType.Raidz1:
+      case TopologyItemType.Raidz:
+      case TopologyItemType.Mirror: {
+        const vdevDisks: TopologyDisk[] = [];
+        for (let i = 0; i < width; i++) {
+          const topologyDisk: unknown = this.generateTopologyDisk();
+          vdevDisks.push(topologyDisk as TopologyDisk);
+        }
+
+        const vdev: unknown = {
```
Comment (undsoft) [inline src/app/core/testing/utils/mock-storage-generator.utils.ts:268]: `const vdev = {} as Vdev.`
`unknown` will swallow errors, like if you mistype one of the fields.

Same in other places in this file.

**Review by undsoft — CHANGES_REQUESTED** [review]
Approach looks good so far. 👍🏻

File: src/app/services/storage.service.ts (hunk 3 above)
Comment (dszidi) [reply src/app/services/storage.service.ts:487]: In those scenarios we specifically compare to the Data vdevs. Case is covered at line 537 where we check for `TopologyWarning.RedundancyMismatch`. We only do this for Metadata and Dedup VDEVs

File: src/app/enums/v-dev-type.enum.ts [hunk 7]
```diff
@@ -1,13 +1,17 @@
...
   Mirror = 'MIRROR',
   Spare = 'SPARE',
   Log = 'LOG',
   Missing = 'MISSING',
   Root = 'ROOT',
   File = 'FILE',
   Raidz = 'RAIDZ',
+  Raidz1 = 'RAIDZ1',
+  Raidz2 = 'RAIDZ2',
+  Raidz3 = 'RAIDZ3',
```
[the remark this answers is not in the document]
Comment (dszidi) [reply src/app/enums/v-dev-type.enum.ts:14]: Yep

`pool.query` via midclt...


image

Actual payload received on Storage Dashboard...


image

**Review by undsoft — APPROVED** [review]
I haven't checked all the specific scenarios, but overall code looks good. Nice work.

**Comment by undsoft** [comment]
backport

**Comment by undsoft** [comment]
backport

**Pull request merged**